torch/_dynamo/backends/train_step.py

```diff
@@ -0,0 +1,111 @@
+    .backward(), optimizer step(), and zero_grad().  It additionally requires no graph-breaks.
+
+    Args:
+        backend_compile_fn (callable): A dynamo compiler function, to be invoked to compile each subgraph.
+    """
+
+    def _compile_fn(mod: fx.GraphModule, fake_inputs: List[torch.Tensor]):
+        """
+        Step 1: Assert inputs (from user) are already Fake, and user their FakeTensorMode
+                (created by dynamo) to fakeify the module's parameters
+        """
+        torch.set_grad_enabled(True)
+        torch._dynamo.utils.assert_no_fake_params_or_buffers(mod)
+        assert len(fake_inputs) > 0, "Expected at least one input"
+        fake_mode = fake_inputs[0].fake_mode
+        assert isinstance(
+            fake_mode, FakeTensorMode
+        ), "Expected a valid FakeTensorMode on dynamo inputs"
+
+        def fakeify_inputs(flat_args):
+            def convert(idx, x):
+                # todo: do we expect symint inputs?
+                assert isinstance(x, torch.Tensor)
+                return fake_mode.from_tensor(x, static_shapes=False)
+
+            return [convert(idx, x) for idx, x in enumerate(flat_args)]
+
+        params = {
+            **dict(mod.named_parameters(remove_duplicate=False)),
+            **dict(mod.named_buffers(remove_duplicate=False)),
+        }
+        params_flat, params_spec = pytree.tree_flatten(params)
+        params_len = len(params_flat)
+        fake_params_flat = fakeify_inputs(params_flat)
+        full_fake_args = fake_params_flat + fake_inputs
+
+        """
+        Step 2: Create a new graphmodule that accepts parameters and user-inputs
+                as inputs to forward(), instead of accessing parameters as attrs.
+        """
+
+        def functional_call(*args, **kwargs):
+            with stateless._reparametrize_module(
+                mod, pytree.tree_unflatten(args[:params_len], params_spec)
+            ):
+                out = mod(*args[params_len:], **kwargs)
+
+            if not isinstance(out, (tuple, list)):
+                raise RuntimeError(
+                    "Graph output must be a tuple() to avoid pytree processing of the ouputs."
+                )
+            return out
+
+        # This fx_g contains expanded backward ops, but also accepts flat params as inputs to forward, so we can't
+        # directly return it
+        assert torch.is_grad_enabled(), "grad isn't enabled before calling make_fx"
+        fx_g = make_fx(functional_call)(*full_fake_args)
+        torch.set_grad_enabled(False)
+
+        """
+        Step 3: Functionalize the resulting flattend graph, producing code with copy_ ops
+                as an epilogue for any inplace/mutating ops such as optimizer update.
+        """
+
+        def retraced_f(*args):
+            return Interpreter(fx_g).run(*args)
+
+        # not really sure why we need inference mode here
```

Comment:
I assume this is because the backward ops are already captured in the fx.Graph, so we don't need to build autograd graph for it any more. 

Besides that, opt will update params inplace. So if this is run under grad mode, autograd engine will complain that we are inplace updating a leaf node.

Reply:
> I assume this is because the backward ops are already captured in the fx.Graph, so we don't need to build autograd graph for it any more.

Yes, this is true for sure.  There isn't a .backwards() call anymore.  (which is why i put in the comment: i thought it probably doesn't matter if we run under grad mode since we aren't actually trying to do anything grad related

> autograd engine will complain that we are inplace updating a leaf node.

This is probably the reason- makes sense now.  cc @bdhirsh

Reply:
ah yep, that's it :)